File: ambeth/Ambeth.Util/ambeth/util/MethodKeyOfType.cs
using System;
using System.Text;

namespace De.Osthus.Ambeth.Util
{
    public class MethodKeyOfType : IPrintable
    {
        protected readonly String methodName;

        protected readonly NewType[] parameterTypes;

        protected readonly NewType returnType;

        public MethodKeyOfType(NewType returnType, String methodName, NewType[] parameterTypes)
        {
            this.returnType = returnType;
            this.methodName = methodName;
            this.parameterTypes = parameterTypes;
        }

        public NewType ReturnType
        {
            get
            {
                return returnType;
            }
        }

        public String MethodName
        {
            get
            {
                return methodName;
            }
        }

        public NewType[] ParameterTypes
        {
            get
            {
                return parameterTypes;
            }
        }

        public override int GetHashCode()
        {
            return returnType.GetHashCode() ^ methodName.GetHashCode() ^ parameterTypes.Length;
        }

        public override bool Equals(Object obj)
        {
            if (Object.ReferenceEquals(this, obj))
            {
                return true;
            }
            if (obj == null)
            {
                return false;
            }
            if (GetType() != obj.GetType())
            {
                return false;
            }
            MethodKeyOfType other = (MethodKeyOfType)obj;
            if (!Object.Equals(methodName, other.methodName))
            {
                return false;
            }
            if (!Arrays.Equals(parameterTypes, other.parameterTypes))
            {
                return false;
            }
            if (!Object.Equals(returnType, other.returnType))
            {
                return false;
            }
            return true;
        }

	    public override String ToString()
	    {
		    StringBuilder sb = new StringBuilder();
		    ToString(sb);
		    return sb.ToString();
	    }

	    public void ToString(StringBuilder sb)
	    {
            sb.Append("MethodKey: ").Append(returnType.ClassName).Append(' ').Append(methodName).Append('(');
		    for (int a = 0, size = parameterTypes.Length; a < size; a++)
		    {
			    if (a > 0)
			    {
                    sb.Append(", ");
			    }
                sb.Append(parameterTypes[a].ClassName);
		    }
            sb.Append(')');
	    }
    }
}

File: jambeth/jambeth-util/src/main/java/de/osthus/ambeth/util/MethodKeyOfType.java
package de.osthus.ambeth.util;

import java.util.Arrays;

import de.osthus.ambeth.repackaged.org.objectweb.asm.Type;

public class MethodKeyOfType
{
	protected final String methodName;

	protected final Type[] parameterTypes;

	public MethodKeyOfType(String methodName, Type[] parameterTypes)
	{
		this.methodName = methodName;
		this.parameterTypes = parameterTypes;
	}

	public String getMethodName()
	{
		return methodName;
	}

	public Type[] getParameterTypes()
	{
		return parameterTypes;
	}

	@Override
	public int hashCode()
	{
		return methodName.hashCode() ^ parameterTypes.length;
	}

	@Override
	public boolean equals(Object obj)
	{
		if (this == obj)
		{
			return true;
		}
		if (obj == null)
		{
			return false;
		}
		if (getClass() != obj.getClass())
		{
			return false;
		}
		MethodKeyOfType other = (MethodKeyOfType) obj;
		if (!EqualsUtil.equals(methodName, other.methodName))
		{
			return false;
		}
		if (!Arrays.equals(parameterTypes, other.parameterTypes))
		{
			return false;
		}
		return true;
	}
}
